File: Python/secant.py
from sympy import *

x = symbols('x')


def secant(expr, x0, x1, tol, maxIter):
    """ Secant Method

    Arguments:
        expr {string} -- is the polinomial expression
        x0 {float} -- is the initial value x_0
        x1 {float} -- is the initial value x_1
        tol {float} -- is the tolerance
        maxIter {int} -- is the max amount of iterations
    Returns:
        xn {float} -- is the solution
        err {float} -- is the error
        _iter {int} -- is the amount of iterations
        fx {float} -- is f(xn)

    """
    errReturn = [0, 0, 0, 0]
    try:
        f = sympify(expr)
        xLast = sympify(x0)
        xn = x1
        for _iter in range(1, maxIter):
            div = f.subs(x, xn) - f.subs(x, xLast)
            if (div == 0):
                print("Error: Division by zero")
                return errReturn
            fx = f.subs(x, xn)
            xNext = xn - (fx*(xn - xLast))/div
            err = abs(xNext - xn)/abs(xNext)
            if (err <= tol):
                break
            xLast = xn
            xn = xNext
        return [N(xn), N(err), _iter, N(fx)]
    except:
        print("There was an error.")
        return errReturn

File: Octave/secant.m
function [xn, err, iter, fx]= secant(f,x0,x1,tol,maxIter)
  f1 = matlabFunction(sym(f));  % Se obtiene la funcion
  xLast = x0;
  xn = x1;
  figure
  hold on
  for iter=0:maxIter
   div = f1(xn) - f(xLast);
    if div == 0
      disp("Error: Division by zero");
      return
    endif
    fx = f1(xn);
    xNext = xn - (fx*(xn - xLast))/div;
    err = abs(xNext - xn) / abs(xNext);
    if err <= tol
      return
    endif
    xLast = xn;
    xn = xNext;
    plot(iter, err, 'ro');
  endfor
endfunction
[xn, err, iter]=secant('exp(x)-(1/x)',0,1,10^-4,50)